### PR TITLE
Move mission dialogue to a json

### PR DESCRIPTION
**Pull request by Coolthulhu**
Moved a gargantuan block (2000 lines of mostly `if/else` trees) to json.

The changes look huge on the outside, but they are mostly just translation from hardcode to json and thus don't need to be reviewed more closely than just by skimming.

But still, I'm not sure if the structure is right.
Now that missiondef.json has the dialogue, it is huge. Dialogue also doesn't blend too well with all the important stuff.
It may be better to keep the dialogue in separate file. Or possibly to keep each mission in a separate file, but that would give us a ton of small files.

Also, I kept the dialogue as a map. It may be cleaner to split it into 8 pairs instead, since those 8 pairs are enforced anyway.

Regardless, this PR is a major unfucking of one of the ugliest sections of the code. It can be improved, but it is an undeniable improvement over master branch without that. Changing one json structure to the other is much easier than translating `if/else` trees to sane data structure.

I'm not sure if after this PR missions will be fully jsonizable, but they might actually be. We won't know until some modder tries modding in a mission and fails because of some shortcoming of the system.

### CODE REVIEW

**Comment by Zireael07** [comment]
> It may be better to keep the dialogue in separate file. 

I think it's a better idea than stuffing it in missiondef.json. Maybe have missiondef and missiondialogue jsons?

**Comment by Coolthulhu** [comment]
I'm trying to change it to look like that, will push changes soon.

**Comment by codemime** [comment]
> Also, I kept the dialogue as a map. It may be cleaner to split it into 8 pairs instead, since those 8 pairs are enforced anyway.

IMHO map is the right design choice here, it's more flexible. Did you mean nine pairs?

> Now that missiondef.json has the dialogue, it is huge.

Can be shortened. From:
```json
[
    "TALK_MISSION_FAILURE",
    "What a shame, that data could have proved invaluable..."
],
```
to (4 times shorter):
```json
"failure": "What a shame, that data could have proved invaluable...",
```
> Or possibly to keep each mission in a separate file, but that would give us a ton of small files.

Against. Apart from some inconvenience, that would possibly increase loading times.

**Comment by Coolthulhu** [comment]
> Can be shortened.

I'm actually doing it the exact way you've described right now. `jq` is very helpful here.

> Against. Apart from some inconvenience, that would possibly increase loading times.

Loading time increase wouldn't be noticeable here, but readability would be greatly improved.
3000 line files are hard to operate on and mixing important data with user strings isn't the best way to do it.

**Comment by codemime** [comment]
What if we split missions into groups (one group - one file, just like for the other stuff)? This way we would get a moderate number of readable files.

> mixing important data with user strings isn't the best way to do it

I think dividing missions and their strings isn't the best way either. They're in a one-to-one relationship. If you edit one thing, you would probably need to adjust the other.

**Comment by Coolthulhu** [comment]
OK, merged back into one block. The file is now pretty huge (2000 lines), but it's much better to have those 2000 lines in a json than in source.

**Comment by vache** [comment]
I will cast my vote for having mission chains in their own files, rather than individual missions.

**Comment by Noctifer-de-Mortem** [comment]
I have some custom mission in mind I could do with this, I will report on the results if this goes through.

**Comment by mugling** [comment]
> Or possibly to keep each mission in a separate file, but that would give us a ton of small files.
> 
> Against. Apart from some inconvenience, that would possibly increase loading times.

I actually tested that a while back. Splitting files makes no measurable difference. The expensive stage is the finalization.

> jq is very helpful here.

jq plus linting is incredibly productive and will allow us to change the format at a later date without too much pain

> OK, merged back into one block. The file is now pretty huge (2000 lines), but it's much better to have those 2000 lines in a json than in source.

I don't want to delay your PR over that as it is indeed an otherwise welcome leap forward. I'd prefer individual files. Filesystems are designed precisely with hierarchies in mind so I see little harm in exploiting that.

File: tools/format/format.conf [hunk 1]
```diff
@@ -39,6 +39,37 @@
 ^snippet:category
 ^snippet:text=ARRAY
 
+# Missions
```
Comment (mugling) [inline tools/format/format.conf]: You need to add a unit test to `tools/format/examples`

Comment (mugling) [inline tools/format/format.conf]: Duplicates part of `format.conf`

**Pull request merged**